pkg/ck8s/workload_cluster.go

```diff
@@ -230,11 +230,18 @@ func (w *Workload) requestJoinToken(ctx context.Context, name string, worker boo
 }
 
 func (w *Workload) RemoveMachineFromCluster(ctx context.Context, machine *clusterv1.Machine) error {
-	request := &apiv1.RemoveNodeRequest{Name: machine.Name, Force: true}
+	if machine == nil {
+		return fmt.Errorf("machine is nil")
+	} else if machine.Status.NodeRef == nil {
+		return fmt.Errorf("machine %s has no node reference", machine.Name)
+	}
+
+	nodeName := machine.Status.NodeRef.Name
```

Comment:
Nice work finding the root cause. I only have one question, otherwise this looks good. Is Status.NodeRef set automatically by CAPI's Machine controller?

Reply:
I think it does:
- https://cluster-api.sigs.k8s.io/developer/architecture/controllers/machine-pool#:~:text=Setting%20NodeRefs%20on%20MachinePool%20instances%20to%20be%20able%20to%20associate%20them%20with%20Kubernetes%20nodes
- https://github.com/kubernetes-sigs/cluster-api/blob/main/internal/controllers/machine/machine_controller_noderef.go#L90-L100

Reply:
yes, this is part of the CAPI contract.